akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -41,119 +39,57 @@ final private[stream] class OutputStreamSourceStage(writeTimeout: FiniteDuration
 
     require(maxBuffer > 0, "Buffer size must be greater than 0")
 
-    val dataQueue = new LinkedBlockingQueue[ByteString](maxBuffer)
+    // Semaphore counting the number of elements we are ready to accept,
+    // which is the demand plus the size of the buffer.
+    val semaphore = new Semaphore(maxBuffer, /* fair =*/ true)
+
     val downstreamStatus = new AtomicReference[DownstreamStatus](Ok)
 
     final class OutputStreamSourceLogic extends GraphStageLogic(shape) {
 
-      var flush: Option[Promise[Unit]] = None
-      var close: Option[Promise[Unit]] = None
-
-      private var dispatcher: ExecutionContext = null // set in preStart
-      private val blockingThreadRef: AtomicReference[Thread] = new AtomicReference() // for postStop interrupt
-
-      private val downstreamCallback: AsyncCallback[Try[ByteString]] =
-        getAsyncCallback {
-          case Success(elem) ⇒ onPush(elem)
-          case Failure(ex)   ⇒ failStage(ex)
-        }
-
-      private val upstreamCallback: AsyncCallback[(AdapterToStageMessage, Promise[Unit])] =
+      private val upstreamCallback: AsyncCallback[AdapterToStageMessage] =
         getAsyncCallback(onAsyncMessage)
 
-      def wakeUp(msg: AdapterToStageMessage): Future[Unit] = {
-        val p = Promise[Unit]()
-        upstreamCallback.invoke((msg, p))
-        p.future
-      }
+      def wakeUp(msg: AdapterToStageMessage): Unit =
+        upstreamCallback.invoke(msg)
 
-      private def onAsyncMessage(event: (AdapterToStageMessage, Promise[Unit])): Unit =
-        event._1 match {
-          case Flush ⇒
-            flush = Some(event._2)
-            sendResponseIfNeed()
+      private def onAsyncMessage(event: AdapterToStageMessage): Unit = {
+        event match {
+          case Send(data) ⇒
+            emit(out, data)
```

Comment:
Wait wat, how do we know we can just emit here?

Reply:
if I understand correctly, unlike `push`, `emit` doesn't need there to be demand first, and will 'buffer' messages by 'stacking' OutHandlers https://doc.akka.io/docs/akka/current/stream/stream-customize.html

Of course we want to avoid unbounded growth of the number of elements buffered in `emit`, but that's what the semaphore takes care of: it makes sure there are a maximum of `maxBuffer` `Send` messages accepted beyond what there's demand for.

Reply:
Yeah, we just use it seldom enough internally that I had forgotten about the semantics